File: consensus/src/main/scala/co/topl/consensus/algebras/BlockHeaderValidationAlgebra.scala
package co.topl.consensus.algebras

import co.topl.consensus.models.{BlockHeader, BlockHeaderValidationFailure, SlotData}

trait BlockHeaderValidationAlgebra[F[_]] {

  /**
   * Indicates if the claimed child is a valid descendent of the parent
   */
  def validate(header: BlockHeader): F[Either[BlockHeaderValidationFailure, BlockHeader]]

  def couldBeValidated(header: BlockHeader, currentHead: SlotData): F[Boolean]
}

File: dart/consensus/lib/algebras/block_header_validation_algebra.dart
import 'package:fpdart/fpdart.dart';
import 'package:topl_protobuf/consensus/models/block_header.pb.dart';

abstract class BlockHeadervalidationAlgebra {
  /**
   * Indicates if the claimed child is a valid descendent of the parent
   */
  Future<List<String>> validate(BlockHeader header);
}
